### PR TITLE
Fix several issues found in the attestcoin pr

### PR DESCRIPTION
**Pull request by BradleyOlson64**
Issues addressed:

1. sudo_account() always returns None — asset owner/freezer silently become the precompile. runtime/src/migrations.rs:186

pallet_sudo::Key is StorageValue<_, T::AccountId, OptionQuery> — the raw value is a bare 32-byte AccountId, with None represented by key absence, not by a SCALE Option prefix. Decoding it as Option<AccountId> reads the account's first byte as the enum discriminant, so it either fails to decode or decodes as None; either way get_storage_value yields None. So migrations.rs:322 falls back to precompile.clone(), and force_asset_status sets owner and freezer to the precompile account — an address nobody holds a key for. Worse, the idempotency check then compares owner == governance where governance is also the precompile, so the migration reports "roles already correct" forever and never self-heals. Fix:

fn sudo_account() -> Option<AccountId> {
    migration::get_storage_value::<AccountId>(b"Sudo", b"Key", &[])
}

2. pallet-assets is configured for permissionless, zero-deposit asset creation. runtime/src/lib.rs:411

CreateOrigin = AsEnsureOriginWithArg<EnsureSigned<AccountId>> with AssetDeposit, MetadataDepositBase, MetadataDepositPerByte, ApprovalDeposit, and AssetAccountDeposit all ConstU128<0>, plus next_asset_id: Some(2) in both chain specs. Any signed account can create unbounded assets and 50-byte metadata entries for tx fees alone — unbacked state bloat on a public chain. Nothing in this PR needs user-created assets; attest-coin is created at genesis or by migration. Recommend CreateOrigin = AsEnsureOriginWithArg<EnsureNever<AccountId>> (or root-only), or non-zero deposits.

3. The cc3-indexer tests were neutered, not fixed. handleEventUnbonded.test.ts, handleEventWithdrawn.test.ts, handleEventBonded.test.ts

Every changed assertion is now a tautology: expect(nodes.length).toBeGreaterThanOrEqual(0), expect(BigInt(node.amount)).toBeGreaterThanOrEqual(0n), and foundMatch only checked if (nodes.length > 0). The suites pass with zero indexed data. The same three handlers were also added to EXCLUDED_HANDLERS in check-cc3-indexer-handlers.sh, so coverage is gone from both directions — nothing would now catch a broken Unbonded/Withdrawn handler.

This also contradicts the PR description, which says these tests "now call force_mint_bond_asset + setMinBondRequirement(non-zero)". Neither force_mint_bond_asset nor force_accrue exists anywhere in the tree — the description documents an approach that isn't in the code. Either land those root helpers as described, or have the tests do the sudo(setMinBondRequirement(non-zero)) + depositTo dance themselves; leaving vacuous assertions is the worst of the three.

4. erc20_balance_of grabs bytes from return without checking for correct length

The actual fix is to stop guessing and reject the ambiguity:

if ret.len() != 32 {
    return Err(PrecompileFailure::Revert {
        exit_status: ExitRevert::Reverted,
        output: b"balanceOf: bad return".to_vec(),
    });
}
Ok(U256::from_big_endian(&ret[..32]))

5. bond_extra can under-collateralize a second attestor. pallets/attestation/src/impls.rs:646 computes extra = bond.min(liquid) and never checks the result reached bond. So the AlreadyBonded → bond_extra path in register_attestor succeeds even when the stash has less liquid attest-coin than min_bond_requirement. You added the aggregate required_bond_for_stash guard on the unbond path but not the register path — the same invariant should be enforced there, or bond_extra should ensure!(liquid >= bond, InsufficientBalance).

6. Cursor bugbot issue:
The new unregister solvency guard rejects exits after MinBondRequirement rises for multi-attestor stashes, but there is no dispatchable to top up active collateral. bond_extra only runs during another registration, which also raises the required aggregate, so operators cannot add funds and leave once the requirement increases.

### CODE REVIEW

File: pallets/attestation/src/lib.rs [hunk 1]
```diff
@@ -1446,6 +1447,28 @@ pub mod pallet {
...
+        ///
+        /// Fails with `NotStash` if the caller has no ledger (nothing to top up — a stash with no
+        /// registered attestor exits via `withdraw_unbonded`) and with `InsufficientBalance` if the
+        /// caller's liquid attest-coin balance is below `amount`.
+        #[pallet::call_index(30)]
+        #[pallet::weight(<T as Config>::WeightInfo::bond_extra())]
+        pub fn bond_extra(origin: OriginFor<T>, amount: BalanceOf<T>) -> DispatchResult {
+            let stash = ensure_signed(origin)?;
+
+            Self::do_bond_extra(&stash, amount)
```
[the remark this answers is not in the document]
Comment (BradleyOlson64) [reply pallets/attestation/src/lib.rs:1480]: Addressed

**Review by DylanVerstraete — COMMENTED** [review]
Reviewed against `origin/attest_coin` (merge-base is its tip, `51602203`, so the branch is current). All six items are addressed in substance, and the two new dispatchables are a genuine improvement on what was asked for. Two things block: one test change that will fail, and the fact that CI has never run on the head commit.

Verification I ran locally on `0fb04b9b`:

- `cargo test -p pallet-attestation --lib` — **270 passed, 0 failed**, including all 8 new tests
- `cargo test -p pallet-evm-precompile-attest-coin --lib` — **40 passed, 0 failed**, including the new `balanceOf` length case
- `cargo fmt --all -- --check` — clean

## Blocking

### 1. `handleEventBonded.test.ts` will fail — the loop-wide amount assertion is unsound

`register_attestor`'s first-bond path emits `Bonded` unconditionally, whatever the requirement is (`pallets/attestation/src/impls.rs`):

```rust
Ok(()) => {
    frame_system::Pallet::<T>::inc_consumers(&stash)?;
    Self::deposit_event(Event::<T>::Bonded { stash, amount: Self::min_bond_requirement(chain_key) });
}
```

With `DefaultMinBondRequirement = 0`, every fresh stash ledger anywhere in the run produces a `Bonded { amount: 0 }` row. Eight other cc3-indexer suites call `registerAttestor`, and four of them are both larger and slower than `handleEventBonded` — `handleEventAttestorChilled` (8.3K), `handleEventAttestorRegistered` (7.4K), `handleEventAttestorUnregistered` (5.9K), `handleEventAttestorActivated` (5.8K) — so under `--runInBand` with jest's default sequencer they run first either way (size-descending with no cache, duration-descending with one). Their zero-amount rows sit inside the `last: 10` window this test iterates, and only three non-zero `Bonded` rows exist in the entire run (the three suites touched here), nowhere near enough to push them out. So this line hits a `0` node:

```ts
for (const node of response.data.bondeds.nodes) {
    expect(BigInt(node.amount)).toBeGreaterThan(0n);   // <-- fails on another suite's Bonded{0}
```

Fix: move the amount assertions inside the `node.stashId === stash.address` branch, where the exact `toEqual(TEST_BOND)` already lives, and keep only the structural assertions (`stashId`, `whoId`, `date`, `blockNumber`) loop-wide.

Note that the same pattern *is* currently safe in the Unbonded and Withdrawn suites, but only incidentally: those events are suppressed at zero value (`if !value.is_zero()` in `remove_attestor_and_emit_event`, `if new_total < old_total` in `do_withdraw_unbonded`). That's the only thing making those two files correct, so it's worth a comment — otherwise a later change that starts emitting zero-value `Unbonded` breaks them silently.

### 2. CI has never run on the head commit

`0fb04b9b` has exactly one check: Cursor Bugbot. The 8 failures on `5b83dc43` look stale rather than real — 7 are `Metadata(IncompatibleCodegen)` from the pre-regen `metadata.scale` (fixed by the metadata commit that followed), and the eighth is the known WS-connect flake (`validatorCycle.test.ts` blowing a 20s `beforeAll` hook, unrelated to this PR). But that needs demonstrating, not assuming. Please re-run and watch two jobs in particular: `cc3-indexer-testing`, which is the only thing that would catch finding 1, and `attestor-compatibility`, since two new call indices change the pallet's metadata hash against a ≥3.130.0 node.

## Worth addressing

### 3. `unbond_surplus` is permissionless with unbounded cost

`required_bond_for_stash` runs `Attestors::iter_prefix(chain_key)` for *every* supported chain. The weights comment concedes this ("the generated weight will need a component for it") but ships `reads(107)` as a guess. The same guard already exists on the unregister path, but that path requires owning an attestor being removed; `bond_extra`/`unbond_surplus` only require a ledger to exist, so this is cheaper to hammer. Either add the benchmark component, or index attestors by stash so the lookup is O(attestors-per-stash) instead of O(registry).

### 4. Land real benchmark numbers

The PR adds working `#[benchmark]` fns *and* placeholder weights marked NOT BENCHMARK-GENERATED with a "regenerate before release" note that nothing enforces. Since the benchmarks exist and pass, `.github/bench.sh -p attestation -b` closes this now rather than leaving a TODO that only a human will remember.

### 5. The sudo-absent fallback keeps the second half of the original bug

The decode fix is right — `pallet_sudo::Key` is `OptionQuery` over a bare `AccountId`, so `get_storage_value::<AccountId>` is correct and the reported failure is genuinely gone. But `sudo_account().unwrap_or_else(|| precompile.clone())` means that with no sudo key, `owner == freezer == governance == precompile`, the idempotency check short-circuits forever, and owner-only calls have no holder. Root `force_asset_status` can still repair it, so this is recoverable rather than critical — but logging an error and skipping the role assignment is more honest than silently assigning roles to an unkeyed account.

### 6. The two new sudo helpers swallow inner dispatch errors

`mintAttestCoin` (via `sudo.sudoAs`) and `setMinBondRequirement` (via `sudo.sudo`) both check only the *outer* extrinsic's `TxStatus.ok`. A failing inner call reports success, and the failure resurfaces later as a confusing `InsufficientBalance` from `register_attestor`. Assert on the `Sudid`/`SudoAsDone` result — or just read the value back, which `setMinBondRequirement` is already positioned to do since it queries `minBondRequirement` anyway.

### 7. Nits

- `next_asset_id: Some(2)` in both chain specs is dead config now that `create` is unreachable.
- `handleEventMinBondRequirementUpdated.test.ts` asserts `minBondRequirementUpdateds(last: 1)` has `chainKey == chain_Anvil1_Key`. The three suites here now emit six extra events for `chain_Anvil2_Key`, and it's only safe because that file (3.0K) sorts last. A `chainKey` filter on that query would remove the invisible dependency on relative file size.

## Verified correct

- **`CreateOrigin = frame_system::EnsureNever<AccountId>` compiles** — frame_system provides an unconditional `EnsureOriginWithArg` impl via `impl_ensure_origin_with_arg_ignoring_arg!`, so `AsEnsureOriginWithArg` isn't needed. Dropping that import and `ConstU128` is clean; no other uses remain in `runtime/src/lib.rs`.
- **The non-zero deposits break nothing.** Nothing in the tree calls signed `assets.create`, `approve_transfer`, `set_metadata`, or `touch`; genesis and `force_create`/`force_set_metadata` bypass deposits; `do_mint` passes `maybe_deposit: None`, so the precompile mint path is unaffected as the comment claims. Magnitudes line up with the rest of the runtime (`deposit(1, 0)` = 20 CTC).
- **`erc20_balance_of`** — behaviourally identical for conforming tokens, rejects the ambiguity. The test is the best one in the PR: parameterizing over `[32, 31, 64, 96]` with `0xEE` filler *and* asserting the conforming length fails later at the transfer (`b"erc20 transfer returned false"`) is what proves the guard doesn't over-fire, which a plain "reverts on bad length" test would have missed.
- **`do_bond_extra`'s `ensure!(liquid >= amount)` is exact**, not off by `min_balance`, because `sync_pool_delta` transfers with `Preservation::Expendable`.
- **`do_unbond_surplus`** — `total_staked` is deliberately unchanged so `sync_pool_delta` sees a zero delta and moves nothing; the comment is accurate. The requirement check is correctly placed *after* the dust sweep, and the swept dust is included in the `Unbonded` amount.
- **Item 5 is genuinely fixed**: the `bond.min(liquid)` clamp is gone, and the zero-bond early return in `bond_extra_for_registration` additionally stops the spurious `Bonded{0}` on second-and-later registrations.
- `call_index` 30/31 are free; `WeightInfo` has exactly one implementor (`weights::WeightInfo<T>`, shared by runtime and mock), so the two trait additions can't miss one. Metadata artifacts are regenerated consistently with `bondExtra`/`unbondSurplus` in `augment-api-tx.ts`.

One thing worth calling out beyond the six items: `unbond_surplus_recovers_an_overshot_top_up_and_reaps_the_stash` and `unbond_surplus_recovers_collateral_after_a_min_bond_decrease` pin a **pre-existing** fund-stranding leak that item 6 only half-identified. Bugbot flagged the no-top-up trap; the inverse leak — an overshot bond or a `MinBondRequirement` *decrease* leaving collateral permanently unreachable, since `unregister_attestor` releases only the current min and `kill_stash` only fires below ED — predates the `bond_extra` dispatchable entirely. Good catch, and good that it's regression-tested.

**Comment by BradleyOlson64** [comment]
Manually reviewed this and removed unnecessary storage bloat + pallet assets config

@DylanVerstraete This is ready for another look

**Review by DylanVerstraete — COMMENTED** [review]
Re-reviewed at `b3e93ea5`. Every item from the previous round is addressed, two of them better than I proposed, and the dust-sweep bug you found on your own is a good catch I missed entirely.

Verified locally on `b3e93ea5`:

- `cargo test -p pallet-attestation --lib` — **272 passed, 0 failed** (was 270; both new tests pass)
- `cargo fmt --all -- --check` — clean
- Base is still current: merge-base with `attest_coin` == its tip (`51602203`)

## Previous round: all closed

| Item | Status |
|---|---|
| `handleEventBonded` loop-wide `amount > 0` (High) | Fixed — structural assertions loop-wide, amount assertions in the owned-stash branch |
| Sudo helpers swallowed inner dispatch errors | Fixed — both read state back |
| `sudo_account()` fallback to the precompile | Fixed, and better than I suggested |
| `next_asset_id: Some(2)` dead config | Fixed — `None` in both genesis fns; no other site carries it |
| `unbond_surplus` unbounded cost | Addressed by argument + repriced weights |
| Jest file-size ordering fragility | Fixed — `chainKey` filter |

Two are worth calling out specifically.

**The no-sudo path is better than what I asked for.** I suggested logging and skipping. You kept `issuer`/`admin` enforcement (correct — the precompile's `mint`/`burn` genuinely cannot work without them) and preserved `owner`/`freezer` instead of inventing an unkeyed holder. The extra guard for `owner != freezer` is a case I hadn't considered at all.

One correction on its comment, though: it attributes the constraint to the extrinsic — *"`force_asset_status` writes both from one value"* — but `force_asset_status` takes `owner` and `freezer` as separate arguments. The single-value collapse is in `apply_roles`, whose signature takes one `owner` and passes it to both. So the skip is safe but stricter than necessary; giving `apply_roles` separate `owner`/`freezer` parameters would let that case be handled rather than skipped. Not worth blocking on — just don't leave the comment blaming the wrong function.

**The `Bonded`/`Unbonded`/`Withdrawn` comments are the right call.** Scoping the amount assertions *and* recording that a loop-wide `amount > 0` only holds because of the zero-value event suppression — with an explicit "do not rely on that here" — is exactly the coupling I was worried about. That will save someone a confusing failure later.

## Your dust-sweep fix is right, and it has one more path

Making the sweep conditional on `required.is_zero()` is correct, and the reasoning in the comment is exactly right: the guard reads the post-sweep value, so it cannot distinguish dust-sweeping from real undercollateralization. I missed this in the first review. `dust_sweep_does_not_consume_collateral_needed_by_remaining_attestors` is a good test — raising `min_balance` to make an otherwise-unreachable branch live is the right way to pin it.

But `do_withdraw_unbonded` reaps on the same ED comparison and never consults `required_bond_for_stash`:

```rust
let ed = T::BondFungibles::minimum_balance(T::BondAssetId::get());
if ledger.unlocking.is_empty() && (ledger.active < ed || ledger.active.is_zero()) {
    Self::kill_stash(&ledger.stash)?;
}
```

So in the exact window you now defend — `0 < required < ED` — a stash that still backs a registered attestor gets reaped, returning its collateral while the `Attestors` entry survives.

I confirmed this rather than reasoning about it. Taking your test's setup (pool funded, `min_balance` raised to 1 unit, `MinBondRequirement` 1000, two attestors on one stash), unregistering one leaves `active == required == 1000`, then maturing the chunk and calling `withdraw_unbonded`:

```
PROBE RESULT: ledger_reaped=true attestor_still_registered=true required=1000
```

The ledger is gone, the attestor is still registered, and its requirement is still 1000. After that the stash cannot even unwind — `unregister_attestor` fails `NotStash`.

This matters *because* of your fix, not despite it. Before it, the unconditional sweep zeroed `active` and the guard rejected the unbond, so you never reached a persistent `0 < active < ED` state with attestors registered. Now that state is legitimate and durable, so the withdraw path needs the same guard — otherwise the window is defended against a spurious error but not against silent collateral loss, which is the worse of the two. The fix looks like the same shape: skip `kill_stash` when `required_bond_for_stash(&stash, None)` is non-zero and fall through to `ledger.update()`.

Same caveat as your comment: reachable only via root `force_asset_status` raising `min_balance` above the min bond. But you chose to defend the window rather than declare it unreachable, and I think that's the right call — this just completes it.

## The weights comment is now wrong for this PR

The `unbond_surplus` doc says CI regenerates it:

> CI's BENCHMARKS workflow regenerates this (it triggers on `**weights.rs` / `**benchmarking.rs` changes and commits `Auto-update pallet weights` back to the branch)

`benchmarks.yml` triggers on `pull_request: branches: [main, usc-testnet, usc-dev]`. This PR targets `attest_coin`, so **BENCHMARKS does not run here** — no pass has been scheduled and none will be. The regeneration only happens when `attest_coin` itself is PR'd to `usc-dev`, which is a different review with different weights in the diff. Worth rewording so nobody reads the current numbers as CI-validated. (Related: #1227 fixes that workflow re-triggering itself on its own weights commits.)

The rest of the justification checks out, and I verified each claim rather than taking it:

- `MaxAttestors = 100` in the runtime, and `MaxAttestorsDefault` returns `T::MaxAttestationNodes::get()` — so the per-chain bound is real.
- `MAXIMUM_BLOCK_WEIGHT` is `Weight::from_parts(WEIGHT_MILLISECS_PER_BLOCK * WEIGHT_REF_TIME_PER_MILLIS, u64::MAX)` — proof_size genuinely unmetered, so the inflated `Estimated` costs nothing.
- 813 reads at ParityDb's read cost is single-digit milliseconds against a multi-second ref_time budget, so "a fraction of a percent" holds.

Keeping the scan rather than a per-stash counter is the right trade for a solvency guard, and the reason given — `Attestors` is the single source of truth, a derived index could drift — is the correct one.

## `pallet-assets` deposits: fine, with one asymmetry

Reverting `AssetDeposit` / metadata deposits / `AssetAccountDeposit` to zero and keeping only `ApprovalDeposit` non-zero is defensible. `create` is unreachable under `EnsureNever`, `set_metadata` is owner-only, and the `ExistenceReason` uniformity argument is a real benefit.

One thing to check before settling: the comment justifies zero `AssetAccountDeposit` because `touch` "costs a transaction per entry anyway". That's the same cost bound as `approve_transfer`, which you *did* give a deposit — and the asymmetry you draw for approvals is that `Approvals` is keyed `(asset, owner, delegate)` so one account can vary the third component freely. `Account` is keyed `(asset, who)`, and `touch_other(id, who)` lets a signed caller create an entry for an arbitrary `who`, which is the same unbounded-key shape. I have not checked whether this `pallet_assets` version gates `touch_other` behind the asset's admin/owner or merely behind the deposit — if it's deposit-gated, a zero deposit makes it free and the two cases are symmetric, and the comment's reasoning for treating them differently doesn't hold. Worth a look at `do_touch`'s privilege check before merging on that argument.

## Still not verified by CI

Only `Cursor Bugbot` has run on `b3e93ea5` (it passed). Same gap as last round: the head commit has no build, no test, no `cc3-indexer-testing` — which is the job that would exercise the four indexer-test changes, since those are integration tests against a live node and cannot be run locally. `integration-test-blockchain` and `check-version` are the known expected-reds per #1215's notes.

For what it's worth, on #1216 the `integration-test-blockchain` failure I flagged last week turned out to be purely a stale base — it passed once rebased. So a fresh run here is likely to be cleaner than the last one you saw, and is the remaining thing I would want before merge.

**Comment by DylanVerstraete** [comment]
Short version of what's still outstanding at `b3e93ea5` — one real bug, one wrong claim, one misattributed comment. Plus a retraction of my own.

### 1. `do_withdraw_unbonded` reaps a stash that still backs an attestor — the only functional issue

`impls.rs` (`do_withdraw_unbonded`), the reap condition never consults `required_bond_for_stash`:

```rust
if ledger.unlocking.is_empty() && (ledger.active < ed || ledger.active.is_zero()) {
    Self::kill_stash(&ledger.stash)?;
}
```

Your dust-sweep fix makes a persistent `0 < active < ED` state legitimate, and this destroys it. Confirmed with a throwaway test built from your own setup (pool funded, `min_balance` 1 unit, `MinBondRequirement` 1000, two attestors on one stash, unregister one, mature the chunk, `withdraw_unbonded`):

```
PROBE RESULT: ledger_reaped=true attestor_still_registered=true required=1000
```

Ledger gone, attestor still registered, requirement still 1000 — and the stash can no longer unwind at all, since `unregister_attestor` then fails `NotStash`.

Fix is the same shape as the one you already applied: skip `kill_stash` when `required_bond_for_stash(&stash, None)` is non-zero and fall through to `ledger.update()`. Worth a test alongside `dust_sweep_does_not_consume_collateral_needed_by_remaining_attestors`, since that suite already builds the exact state.

### 2. The `unbond_surplus` weights comment is wrong for this PR

> CI's BENCHMARKS workflow regenerates this (it triggers on `**weights.rs` / `**benchmarking.rs` changes …)

`benchmarks.yml` is `pull_request: branches: [main, usc-testnet, usc-dev]`. This PR targets `attest_coin`, so no BENCHMARKS run has been or will be scheduled here. The numbers are hand-written and will stay that way until `attest_coin` itself is PR'd to `usc-dev`. Just needs rewording so they aren't read as CI-validated — the rest of that justification I checked and it all holds (`MaxAttestors = 100`, proof_size is `u64::MAX` so unmetered, 813 ParityDb reads is a fraction of a percent of the ref_time budget).

### 3. The no-sudo guard's comment blames the wrong function

> `force_asset_status` writes both from one value, so it cannot preserve a split owner/freezer

`force_asset_status` takes `owner` and `freezer` as separate arguments. The collapse is in `apply_roles`, which accepts a single `owner` and passes it to both. The skip is safe, so this is comment-only — though giving `apply_roles` separate `owner`/`freezer` params would let that case be handled instead of skipped.

### Retracting my `touch_other` concern — you were right

I asked you to double-check whether zero `AssetAccountDeposit` opens the same unbounded-key vector as `Approvals`. It doesn't, and I should have checked before raising it. `touch_other` calls `do_touch(id, who, origin, check_depositor = true)`, and `do_touch` gates on:

```rust
ensure!(
    !check_depositor || &depositor == &details.admin || &depositor == &details.freezer,
    Error::<T, I>::NoPermission
);
```

So creating an account for an arbitrary `who` is admin/freezer-only — the precompile and sudo for attest coin, not an arbitrary caller. Plain `touch()` passes `check_depositor = false` but with `who == origin`, so it can only ever create its own `(asset, origin)` entry. Your asymmetry is exactly right: `Approvals` keyed `(asset, owner, delegate)` lets one account vary the third component freely, `Account` keyed `(asset, who)` does not. The pallet-assets config is settled as far as I'm concerned.

### Not a code issue, but still open

Only Cursor Bugbot has run on `b3e93ea5`. No build, no tests, and no `cc3-indexer-testing` — which is the only thing that can exercise the four indexer-test changes, since those are integration tests against a live node. Worth a run before merge; on #1216 the `integration-test-blockchain` failure I flagged turned out to be a stale base and went green once rebased, so this should look cleaner than the last run you saw.

Everything else from the previous round is closed. Verified locally at `b3e93ea5`: `cargo test -p pallet-attestation --lib` 272 passed / 0 failed, `cargo fmt --all --check` clean.
